example/src/test/scala/monocle/HttpRequestExample.scala

```diff
@@ -0,0 +1,91 @@
+package monocle
+
+import monocle.macros.{GenLens, GenIso, GenPrism}
+
+/**
+ * Show how could we use Monocle to handle custom case classes, objects
+ */
+class HttpRequestExample extends MonocleSuite {
+  import HttpRequestExample._
+
+  val r1 = HttpRequest(
+    GET,
+    URI("localhost", 8080, "/ping", Map("hop" -> "5")),
+    Map("socket_timeout" -> "20", "connection_timeout" -> "10"),
+    "")
+
+  val r2 = HttpRequest(
+    POST,
+    URI("gooogle.com", 443, "/search", Map("keyword" -> "monocle")),
+    Map.empty,
+    "")
+
+  val method = GenLens[HttpRequest](_.method)
+  val uri = GenLens[HttpRequest](_.uri)
+  val headers = GenLens[HttpRequest](_.headers)
+  val body = GenLens[HttpRequest](_.body)
+
+  val host = GenLens[URI](_.host)
+  val query = GenLens[URI](_.query)
+
+  val get: Prism[HttpMethod, Unit] = GenPrism[HttpMethod, GET.type] composeIso GenIso.unit[GET.type]
+  val post = GenPrism[HttpMethod, POST.type] composeIso GenIso.unit[POST.type]
+
+  test("get and post") {
+    (method composePrism get).isMatching(r1) shouldBe true
+    (method composePrism post).isMatching(r1) shouldBe false
+    (method composePrism post).getOption(r2) shouldBe Some(())
+  }
+
+  test("host") {
+    (uri composeLens host).set("google.com")(r2) shouldBe
+      r2.copy(uri = r2.uri.copy(host = "google.com"))
+  }
+
+  test("query using index") {
+    val r = (uri
+      composeLens query
+      composeOptional index("hop")
+      composePrism stringToInt).modify(_ + 10)(r1)
+
+    r.uri.query.get("hop") shouldBe Some("15")
+  }
+
+  test("query using at") {
+    /**
+     *  `at` returns Lens[S, Option[A]] while `index` returns Optional[S, A]
```

Comment:
here you could use `composeOption index("hop")`, I am not sure it is what you meant